File: simple_functions/src/test/scala/com/jinloes/simple_functions/TreeTraverserTest.scala
package com.jinloes.simple_functions

import javax.swing.tree.DefaultMutableTreeNode

import com.jinloes.simple_functions.tree.TreeTraverser
import org.junit.runner.RunWith
import org.scalatest.junit.JUnitRunner
import org.scalatest.{FlatSpec, Matchers}

import scala.collection.JavaConverters._

/**
  * Tests for {@link TreeTraverser}
  */
@RunWith(classOf[JUnitRunner])
class TreeTraverserTest extends FlatSpec with Matchers {
  "A tree traverser" should "be able to perform an preorder traversal" in {
    /*
            A
           / \
          B   C
         /
        D
    */
    val root: DefaultMutableTreeNode = new DefaultMutableTreeNode("A")
    val child1: DefaultMutableTreeNode = new DefaultMutableTreeNode("B")
    val child2: DefaultMutableTreeNode = new DefaultMutableTreeNode("C")
    val child1child: DefaultMutableTreeNode = new DefaultMutableTreeNode("D")
    root.insert(child1, 0)
    root.insert(child2, 1)
    child1.insert(child1child, 0)
    val actual = TreeTraverser.preOrder(root)
    actual should be(List("A", "B", "D", "C").asJava)
  }

  it should " be traverse an empty graph" in {
    val root: DefaultMutableTreeNode = null
    val actual = TreeTraverser.preOrder(root)
    actual should be(List().asJava)
  }

  it should "be able to perform an inorder traversal" in {
    /*
          A
         / \
        B   C
       / \
      D   E
     */
    val root: DefaultMutableTreeNode = new DefaultMutableTreeNode("A")
    val child1: DefaultMutableTreeNode = new DefaultMutableTreeNode("B")
    val child2: DefaultMutableTreeNode = new DefaultMutableTreeNode("C")
    val child1child: DefaultMutableTreeNode = new DefaultMutableTreeNode("D")
    val child1child2: DefaultMutableTreeNode = new DefaultMutableTreeNode("E")
    root.insert(child1, 0)
    root.insert(child2, 1)
    child1.insert(child1child, 0)
    child1.insert(child1child2, 1)
    val actual = TreeTraverser.inOrder(root)
    actual should be(List("D", "B", "E", "A", "C").asJava)
  }

  it should "perform an inorder traversal on an empty graph" in {
    val root: DefaultMutableTreeNode = null
    val actual = TreeTraverser.inOrder(root)
    actual should be('empty)
  }

  it should "be able to perform an postorder traversal" in {
    /*
          A
         / \
        B   C
       / \
      D   E
     */
    val root: DefaultMutableTreeNode = new DefaultMutableTreeNode("A")
    val child1: DefaultMutableTreeNode = new DefaultMutableTreeNode("B")
    val child2: DefaultMutableTreeNode = new DefaultMutableTreeNode("C")
    val child1child: DefaultMutableTreeNode = new DefaultMutableTreeNode("D")
    val child1child2: DefaultMutableTreeNode = new DefaultMutableTreeNode("E")
    root.insert(child1, 0)
    root.insert(child2, 1)
    child1.insert(child1child, 0)
    child1.insert(child1child2, 1)
    val actual = TreeTraverser.postOrder(root)
    actual should be(List("D", "E", "B", "C", "A").asJava)
  }
  it should "perform a postorder traversal on an empty graph" in {
    val root: DefaultMutableTreeNode = null
    val actual = TreeTraverser.postOrder(root)
    actual should be('empty)
  }
}

File: simple_functions/src/test/java/com/jinloes/simple_functions/TreeTraverserTest.java
package com.jinloes.simple_functions;

import com.jinloes.simple_functions.tree.TreeNode;
import com.jinloes.simple_functions.tree.TreeTraverser;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

class TreeTraverserTest {

    private TreeNode<String> root;

    @BeforeEach
    void setUp() {
        // Level 0:      A
        // Level 1:    B   C
        // Level 2:   D E
        root = new TreeNode<>("A",
            new TreeNode<>("B", new TreeNode<>("D"), new TreeNode<>("E")),
            new TreeNode<>("C"));
    }

    @Test
    void nullRootReturnsEmpty() {
        assertThat(TreeTraverser.preOrder(null)).isEmpty();
        assertThat(TreeTraverser.inOrder(null)).isEmpty();
        assertThat(TreeTraverser.postOrder(null)).isEmpty();
    }

    @Test
    void preOrder() {
        assertThat(TreeTraverser.preOrder(root)).isEqualTo(List.of("A", "B", "D", "E", "C"));
    }

    @Test
    void singleNode() {
        TreeNode<String> single = new TreeNode<>("A");
        assertThat(TreeTraverser.preOrder(single)).isEqualTo(List.of("A"));
        assertThat(TreeTraverser.inOrder(single)).isEqualTo(List.of("A"));
        assertThat(TreeTraverser.postOrder(single)).isEqualTo(List.of("A"));
    }

    @Test
    void rightSkewedTree() {
        // Level 0:  A
        // Level 1:    B
        // Level 2:      C
        TreeNode<String> skewed = new TreeNode<>("A", null, new TreeNode<>("B", null, new TreeNode<>("C")));
        assertThat(TreeTraverser.preOrder(skewed)).isEqualTo(List.of("A", "B", "C"));
        assertThat(TreeTraverser.inOrder(skewed)).isEqualTo(List.of("A", "B", "C"));
        assertThat(TreeTraverser.postOrder(skewed)).isEqualTo(List.of("C", "B", "A"));
    }

    @Test
    void inOrder() {
        assertThat(TreeTraverser.inOrder(root)).isEqualTo(List.of("D", "B", "E", "A", "C"));
    }

    @Test
    void postOrder() {
        assertThat(TreeTraverser.postOrder(root)).isEqualTo(List.of("D", "E", "B", "C", "A"));
    }
}